photon-ml/src/main/scala/com/linkedin/photon/ml/evaluation/AreaUnderROCCurveLocalEvaluator.scala

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright 2016 LinkedIn Corp. All rights reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linkedin.photon.ml.evaluation
+
+import scala.collection.Map
+
+import com.linkedin.photon.ml.constants.MathConst
+
+
+/**
+ * Local evaluator for binary classification problems
+ *
+ * @param labelAndOffsetAndWeights a (id -> (label, offset, weight)) map
+ * @param defaultScore the default score used to compute the metric
+ */
+protected[ml] class AreaUnderROCCurveLocalEvaluator(
+    labelAndOffsetAndWeights: Map[Long, (Double, Double, Double)],
+    defaultScore: Double = 0.0) extends LocalEvaluator {
+
+  val numLabels = labelAndOffsetAndWeights.size
+
+  /**
+   * Evaluate the scores of the model
+   *
+   * @param scores the scores to evaluate
+   * @return score metric value
+   */
+  override def evaluate(scores: Map[Long, Double]): Double = {
+    val scoresAndLabelAndWeight = new Array[(Double, Double, Double)](numLabels)
+    var pos = 0
+    labelAndOffsetAndWeights.foreach { case (id, (label, offset, weight)) =>
+      val score = scores.getOrElse(id, defaultScore) + offset
+      scoresAndLabelAndWeight(pos) = (score, label, weight)
+      pos += 1
+    }
+    val sortedScoresAndLabelAndWeight = scoresAndLabelAndWeight.sortBy(_._1)(Ordering[Double].reverse)
+
+    var aucAccum = 0.0
+    var positiveCount = 0.0
+    var negativeCount = 0.0
+
+    sortedScoresAndLabelAndWeight.foreach { case (score, label, weight) =>
+      if (label > MathConst.POSITIVE_RESPONSE_THRESHOLD) {
+        positiveCount += weight
+      } else {
+        aucAccum += positiveCount
+        negativeCount += weight
+      }
+    }
+
+    //normalize ACU over the total area
+    aucAccum / (positiveCount * negativeCount)
```

Comment:
@XianXing I believe there's a subtle flaw in this computation. If we have two identical scores with conflicting ground-truth labels, we're accumulating a little too much area for the "true" one. For example, in the following case:

```
labels = [0, 1, 0, 0, 1, 1, 1]
scores = [-0.1, -1, 0, -1, 2, 6, 8]
```

we compute an AUC of **0.8333** here, but it should be **0.7917**.